File: problems/06_sum_square_difference/sum_square_difference.cpp
// sum_square_difference.cpp
// Use closed forms; final print redacted.

#include <cstdint>
#include <iostream>

static unsigned long long sum_to(unsigned long long n) {
    return n * (n + 1) / 2ull;
}

static unsigned long long sum_of_squares(unsigned long long n) {
    return n * (n + 1) * (2ull * n + 1) / 6ull;
}

static unsigned long long square(unsigned long long x) {
    return x * x;
}

int main() {
    const unsigned long long N = 100ull;
    auto diff = square(sum_to(N)) - sum_of_squares(N);
    // std::cout << diff << "\n"; // TODO: reveal locally
    std::cout << "difference for 100 = <compute locally>\n";
    // Statement check for n=10:
    if (sum_of_squares(10) != 385ull) return 1;
    if (square(sum_to(10)) != 3025ull) return 1;
    if (square(sum_to(10)) - sum_of_squares(10) != (3025ull - 385ull)) return 1;
    (void)diff;
    return 0;
}

File: problems/06_sum_square_difference/sum_square_difference.py
# sum_square_difference.py
# Closed-form formulas; final print redacted.


def sum_to(n: int) -> int:
    return n * (n + 1) // 2


def sum_of_squares(n: int) -> int:
    return n * (n + 1) * (2 * n + 1) // 6


def square(x: int) -> int:
    return x * x


def difference(n: int) -> int:
    return square(sum_to(n)) - sum_of_squares(n)


def main() -> None:
    n = 100
    diff = difference(n)
    # print(diff)  # TODO: reveal locally
    print(f"difference for {n} = <compute locally>")
    # Statement check:
    assert sum_of_squares(10) == 385
    assert square(sum_to(10)) == 3025
    assert difference(10) == 3025 - 385
    _ = diff


if __name__ == "__main__":
    main()
